cpp/src/arrow/array/array_nested.cc

```diff
@@ -469,6 +470,49 @@ inline void SetListData(VarLengthListLikeArray<TYPE>* self,
   self->values_ = MakeArray(self->data_->child_data[0]);
 }
 
+Result<std::shared_ptr<Array>> FlattenLogicalListRecursively(const Array& in_array,
+                                                             MemoryPool* memory_pool) {
+  std::shared_ptr<Array> array = in_array.Slice(0, in_array.length());
+  for (auto kind = array->type_id(); is_list(kind) || is_list_view(kind);
```

Comment:
`is_list` and `is_list_view` would exclude `MAP` type.

Reply:
Indeed.